stl/inc/span

```diff
@@ -0,0 +1,762 @@
+    }
+};
+
+// STRUCT TEMPLATE _Span_extent_type
+template <size_t _Extent>
+struct _Span_extent_type {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type([[maybe_unused]] const size_t _Size) noexcept {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Size == _Extent, "Mismatch between static extent and size of initializing data");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept {
+        static_assert(_OtherExtent == _Extent || _OtherExtent == dynamic_extent,
+            "Mismatch between static extent and size of initializing data");
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_OtherExtent == dynamic_extent) {
+            _STL_VERIFY(_Ext.size() == _Extent, "Mismatch between static extent and size of initializing data");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Extent;
+    }
+};
+
+template <>
+struct _Span_extent_type<dynamic_extent> {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext.size()) {}
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+template <class _Ty, size_t _Extent>
+class span;
+
+#ifdef __cpp_lib_concepts
+
+// VARIABLE TEMPLATE _Is_span_v
+template <class>
+inline constexpr bool _Is_span_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_span_v<span<_Ty, _Extent>> = true;
+
+// VARIABLE TEMPLATE _Is_std_array
+template <class>
+inline constexpr bool _Is_std_array_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_std_array_v<array<_Ty, _Extent>> = true;
+
+// clang-format off
+template<class _It, class _Ty>
+concept _Is_span_compatible_iterator = contiguous_iterator<_It>
+                                    && is_convertible_v<remove_reference_t<iter_reference_t<_It>>(*)[], _Ty(*)[]>;
+
+template<class _Sentinel, class _It>
+concept _Is_span_compatible_sentinel = sized_sentinel_for<_Sentinel, _It>
+                                    && !is_convertible_v<_Sentinel, size_t>;
+
+template <class _Range, class _Ty>
+concept _Is_span_compatible_range = !is_array_v<remove_cvref_t<_Range>>
+                                 && !_Is_span_v<remove_cvref_t<_Range>>
+                                 && !_Is_std_array_v<remove_cvref_t<_Range>>
+                                 && _RANGES contiguous_range<_Range>
+                                 && _RANGES sized_range<_Range>
+                                 && (_RANGES forward_range<_Range> || is_const_v<_Ty>)

```

Comment:
This needs to be `_Forwarding_range` rather than `forward_range` until we implement P1870 and it can become `safe_range`.
```suggestion
                                 && (_RANGES _Forwarding_range<_Range> || is_const_v<_Ty>)
```